File: src/test/java/io/beekeeper/bots/pizza/ParserTest.java
package io.beekeeper.bots.pizza;

import org.junit.Assert;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

public class ParserTest {

    @Test
    public void testSomething() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola", "Gorgonzola Whee");
        menu.put("Margherita", "Margherita");
        Parser<String> p = new Parser<String>(menu);

        String result = p.parse("gorgonzola");

        Assert.assertEquals("Gorgonzola Whee", result);
    }

    @Test
    public void testDoesNotMatchIfOnlyOneWord() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola Thing", "Gorgonzola Whee");
        menu.put("Margherita", "Margherita");
        Parser<String> p = new Parser<String>(menu);

        String result = p.parse("gorgonzola");

        Assert.assertEquals(null, result);
    }

    @Test
    public void testMatchesWordsNotAtStart() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola Thing Pizza", "Gorgonzola Whee");
        menu.put("Margherita", "Margherita");
        Parser<String> p = new Parser<>(menu);

        String result = p.parse("thing pizza");

        Assert.assertEquals("Gorgonzola Whee", result);
    }

    @Test
    public void testMatchesMinimalPrefixes() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola Thing Pizza", "Gorgonzola Whee");
        menu.put("Margherita", "Margherita");
        Parser<String> p = new Parser<>(menu);

        String result = p.parse("gor th pi");

        Assert.assertEquals("Gorgonzola Whee", result);
    }

    @Test
    public void testMatchesNaturalString() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola Pizza", "Gorgonzola Whee");
        menu.put("Margherita", "Margherita");
        Parser<String> p = new Parser<>(menu);

        String result = p.parse("One gorgonzola pizza plx!");

        Assert.assertEquals("Gorgonzola Whee", result);
    }

    @Test
    public void testFindsBestWhenMultipleMatch() {
        Map<String, String> menu = new HashMap<>();
        menu.put("Gorgonzola Pizza", "Gorgonzola Whee");
        menu.put("Gorgonzola", "Margherita");
        menu.put("Gorgonzola Pizza Margherita", "correct");
        Parser<String> p = new Parser<>(menu);

        String result = p.parse("gorgonzola pizza margherita");

        Assert.assertEquals("correct", result);
    }
}

File: src/test/java/io/beekeeper/bots/pizza/providers/dieci/ParserTest.kt
package io.beekeeper.bots.pizza.providers.dieci

import io.beekeeper.bots.pizza.test.TestMenuItem
import org.junit.Assert.assertEquals
import org.junit.Test

class ParserTest {

    @Test
    fun testSomething() {
        val menu = mapOf(
                "Gorgonzola" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("gorgonzola")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

    @Test
    fun testMatchesWordsNotAtStart() {
        val menu = mapOf(
                "Gorgonzola Thing Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("thing pizza")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

    @Test
    fun testMatchesMinimalPrefixes() {
        val menu = mapOf(
                "Gorgonzola Thing Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("gor th pi")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

    @Test
    fun testMatchesNaturalString() {
        val menu = mapOf(
                "Gorgonzola Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("One gorgonzola pizza plx!")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

    @Test
    fun testFindsBestWhenMultipleMatch() {
        val menu = mapOf(
                "Gorgonzola Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Gorgonzola" to TestMenuItem("Margherita"),
                "Gorgonzola Pizza Margherita" to TestMenuItem("correct")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("gorgonzola pizza margherita")

        assertEquals("correct", result?.articleName)
    }

    @Test
    fun testFindsWithUmlaut() {
        val menu = mapOf(
                "öGorgonzöla Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("ogorgonzola pizza")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

    @Test
    fun testFindsWithUmlaut2() {
        val menu = mapOf(
                "oGorgonzöla Pizza" to TestMenuItem("Gorgonzola Whee"),
                "Margherita" to TestMenuItem("Margherita")
        )
        val p = DieciMenuItemParser<TestMenuItem>(menu)

        val result = p.parse("ögörgönzölä pizzä")

        assertEquals("Gorgonzola Whee", result?.articleName)
    }

}
